// qiankun/lib/utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sleep = sleep;
exports.isConstructable = isConstructable;
exports.getDefaultTplWrapper = getDefaultTplWrapper;
exports.getWrapperId = getWrapperId;
exports.validateExportLifecycle = validateExportLifecycle;
exports.Deferred = void 0;

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _snakeCase2 = _interopRequireDefault(require("lodash/snakeCase"));

function sleep(ms) {
  return new Promise(function (resolve) {
    return setTimeout(resolve, ms);
  });
}

function isConstructable(fn) {
  var constructableFunctionRegex = /^function\b\s[A-Z].*/;
  var classRegex = /^class\b/; // 有 prototype 并且 prototype 上有定义一系列非 constructor 属性，则可以认为是一个构造函数

  return fn.prototype && Object.getOwnPropertyNames(fn.prototype).filter(function (k) {
    return k !== 'constructor';
  }).length || constructableFunctionRegex.test(fn.toString()) || classRegex.test(fn.toString());
}

function getDefaultTplWrapper(id) {
  return function (tpl) {
    return "<div id=\"" + getWrapperId(id) + "\">" + tpl + "</div>";
  };
}

function getWrapperId(id) {
  return "__qiankun_subapp_wrapper_for_" + (0, _snakeCase2.default)(id) + "__";
}
/** 校验子应用导出的 生命周期 对象是否正确 */


function validateExportLifecycle(exports) {
  var _a = exports !== null && exports !== void 0 ? exports : {},
      bootstrap = _a.bootstrap,
      mount = _a.mount,
      unmount = _a.unmount;

  return (0, _isFunction2.default)(bootstrap) && (0, _isFunction2.default)(mount) && (0, _isFunction2.default)(unmount);
}

var Deferred =
/** @class */
function () {
  function Deferred() {
    var _this = this;

    this.promise = new Promise(function (resolve, reject) {
      _this.resolve = resolve;
      _this.reject = reject;
    });
  }

  return Deferred;
}();

exports.Deferred = Deferred;
