// qiankun/lib/apis.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerMicroApps = registerMicroApps;
exports.loadMicroApp = loadMicroApp;
exports.start = start;
exports.frameworkConfiguration = void 0;

var _tslib = require("tslib");

var _singleSpa = require("single-spa");

var _loader = require("./loader");

var _prefetch = require("./prefetch");

var _utils = require("./utils");

window.__POWERED_BY_QIANKUN__ = true;
var microApps = []; // eslint-disable-next-line import/no-mutable-exports

var frameworkConfiguration = {};
exports.frameworkConfiguration = frameworkConfiguration;
var frameworkStartedDefer = new _utils.Deferred();

function registerMicroApps(apps, lifeCycles) {
  var _this = this; // Each app only needs to be registered once


  var unregisteredApps = apps.filter(function (app) {
    return !microApps.some(function (registeredApp) {
      return registeredApp.name === app.name;
    });
  });
  microApps = (0, _tslib.__spread)(microApps, unregisteredApps);
  unregisteredApps.forEach(function (app) {
    var name = app.name,
        activeRule = app.activeRule,
        props = app.props,
        appConfig = (0, _tslib.__rest)(app, ["name", "activeRule", "props"]);
    (0, _singleSpa.registerApplication)({
      name: name,
      app: function app() {
        return (0, _tslib.__awaiter)(_this, void 0, void 0, function () {
          return (0, _tslib.__generator)(this, function (_a) {
            switch (_a.label) {
              case 0:
                return [4
                /*yield*/
                , frameworkStartedDefer.promise];

              case 1:
                _a.sent();

                return [2
                /*return*/
                , (0, _loader.loadApp)((0, _tslib.__assign)({
                  name: name,
                  props: props
                }, appConfig), frameworkConfiguration, lifeCycles)];
            }
          });
        });
      },
      activeWhen: activeRule,
      customProps: props
    });
  });
}

function loadMicroApp(app, configuration) {
  if (configuration === void 0) {
    configuration = frameworkConfiguration;
  }

  var props = app.props,
      appConfig = (0, _tslib.__rest)(app, ["props"]);
  return (0, _singleSpa.mountRootParcel)(function () {
    return (0, _loader.loadApp)(appConfig, configuration);
  }, (0, _tslib.__assign)({
    domElement: document.createElement('div')
  }, props));
}

function start(opts) {
  if (opts === void 0) {
    opts = {};
  }

  exports.frameworkConfiguration = frameworkConfiguration = (0, _tslib.__assign)({
    prefetch: true,
    singular: true,
    sandbox: true
  }, opts);
  var prefetch = frameworkConfiguration.prefetch,
      sandbox = frameworkConfiguration.sandbox,
      singular = frameworkConfiguration.singular,
      urlRerouteOnly = frameworkConfiguration.urlRerouteOnly,
      importEntryOpts = (0, _tslib.__rest)(frameworkConfiguration, ["prefetch", "sandbox", "singular", "urlRerouteOnly"]);

  if (prefetch) {
    (0, _prefetch.prefetchApps)(microApps, prefetch, importEntryOpts);
  }

  if (sandbox) {
    if (!window.Proxy) {
      console.warn('[qiankun] Miss window.Proxy, proxySandbox will degenerate into snapshotSandbox'); // 快照沙箱不支持非 singular 模式

      if (!singular) {
        console.error('[qiankun] singular is forced to be true when sandbox enable but proxySandbox unavailable');
        frameworkConfiguration.singular = true;
      }
    }
  }

  (0, _singleSpa.start)({
    urlRerouteOnly: urlRerouteOnly
  });
  frameworkStartedDefer.resolve();
}
